aws/resource_aws_sqs_queue_policy.go

```diff
@@ -40,18 +43,51 @@ func resourceAwsSqsQueuePolicy() *schema.Resource {
 
 func resourceAwsSqsQueuePolicyUpsert(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).sqsconn
+	policy := d.Get("policy").(string)
 	url := d.Get("queue_url").(string)
 
 	_, err := conn.SetQueueAttributes(&sqs.SetQueueAttributesInput{
 		QueueUrl: aws.String(url),
 		Attributes: aws.StringMap(map[string]string{
-			"Policy": d.Get("policy").(string),
+			"Policy": policy,
 		}),
 	})
 	if err != nil {
 		return fmt.Errorf("Error updating SQS attributes: %s", err)
 	}
 
+	// https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SetQueueAttributes.html
+	// When you change a queue's attributes, the change can take up to 60 seconds
+	// for most of the attributes to propagate throughout the Amazon SQS system.
+	wait := resource.StateChangeConf{
+		Pending:    []string{""},
+		Target:     []string{"SQS queue policy updated"},
+		Timeout:    1 * time.Minute,
+		MinTimeout: 1 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			out, err := conn.GetQueueAttributes(&sqs.GetQueueAttributesInput{
+				QueueUrl:       aws.String(url),
+				AttributeNames: []*string{aws.String("Policy")},
+			})
+			if err != nil {
+				return out, "", err
+			}
+			queuePolicy, ok := out.Attributes["Policy"]
+			if ok {
+				equivalent, err := awspolicy.PoliciesAreEquivalent(*queuePolicy, policy)
+				if err != nil || !equivalent {
+					return out, "", nil
+				}
+				return out, "SQS queue policy updated", nil
+			}
+			return out, "", nil
+		},
+	}
+	_, err = wait.WaitForState()
```

Comment:
Have you considered using the `resource.Retry` helper instead of this wrapper? We'd still need to make up the error message when policies aren't equivalent, but I feel that it'd be a little bit cleaner as the whole thing isn't really "stateful". 🤔

Reply:
Also we're catching _all errors_ here and treating them as "retryable" which (IMO) isn't ideal. I assume the goal is to only catch `AWS.SimpleQueueService.NonExistentQueue`?

Reply:
Good call on both of these -- I'll get this PR updated shortly.